### PR TITLE
EOS-24862 : Implement and test object range read

### PR DESCRIPTION
**Pull request by priyanka25081999**
- Experiment object range read function 
- Implement a class for range read 
- Add a test file for developed class 
- Update makefile and config files
- Remove temporary object tagset file 

Signed-off-by: Priyanka Salunke <priyanka.s.salunke@seagate.com>

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: s3/replication/common/src/s3replicationcommon/s3_put_object_tagging.py [hunk 1]
```diff
@@ -127,6 +127,9 @@ async def send(self):
                     self._logger.info('Response headers {}'.format(
                         self._response_headers))
 
+                    # Delete temporary tagset file.
+                    os.system('rm -rf tagset.xml')
```
Comment (mbcortx) [inline s3/replication/common/src/s3replicationcommon/s3_put_object_tagging.py:131]: I may have missed that we are using a temporary file to prepare the tagset. We should find another way.

File: s3/replication/common/src/s3replicationcommon/aws_v4_signer.py [hunk 2]
```diff
@@ -195,7 +195,8 @@ def prepare_signed_header(
             http_request,
             request_uri,
             query_params,
-            body):
+            body,
+            obj_range=None):
```
Comment (mbcortx) [inline s3/replication/common/src/s3replicationcommon/aws_v4_signer.py:199]: for header signature, adding range is given

File: s3/replication/common/src/s3replicationcommon/s3_range_read.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+import aiohttp
+import sys
+from s3replicationcommon.aws_v4_signer import AWSV4Signer
+from s3replicationcommon.log import fmt_reqid_log
+from s3replicationcommon.s3_common import S3RequestState
+from s3replicationcommon.timer import Timer
+
+
+class S3AsyncGetRangeRead:
```
Comment (mbcortx) [inline s3/replication/common/src/s3replicationcommon/s3_range_read.py:28]: I did not get it here. Why do we need another GetRangeRead? We already have GetObject.  

GetObject(BucketName, ObjectName, offset=0, length=0)

range is offset,offset+length

File: s3/replication/common/src/s3replicationcommon/s3_range_read.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+    def get_state(self):
+        """Returns current request state."""
+        return self._state
+
+    def get_execution_time(self):
+        """Return total time for GET Object operation."""
+        return self._timer.elapsed_time_ms()
+
+    def get_etag(self):
```
Comment (mbcortx) [inline s3/replication/common/src/s3replicationcommon/s3_range_read.py:57]: those are same as GetObject

File: s3/replication/common/src/s3replicationcommon/s3_range_read.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+                    self._state = S3RequestState.FAILED
+                    self._logger.error(
+                        fmt_reqid_log(self._request_id) +
+                        'GET Object failed with http status: {}'.
+                        format(resp.status) +
+                        '\nError Response: Invalid object range')
+                    return
+
+                while True:
+                    resp_data = await resp.content.read()
```
Comment (mbcortx) [inline s3/replication/common/src/s3replicationcommon/s3_range_read.py:122]: we read in chunks in GetObject. Range read can also read in chunks. I think we can combine those:
get object and return everything,  read in chunks (default chunk size can be zero),  read range if given

File: s3/replication/common/src/s3replicationcommon/s3_range_read.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                    if not resp_data:
+                        break
+                    yield resp_data
```
Comment (mbcortx) [inline s3/replication/common/src/s3replicationcommon/s3_range_read.py:125]: do we need a generator here? we do not read in chunks, it is read() not read(length)

File: s3/replication/common/tests/system/test_range_read.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    bucket_name = config.source_bucket_name
+
+    # Generate object names
+    object_name = config.object_name_prefix + "test"
+    request_id = "dummy-request-id"
+
+    # Get object range from config
+    object_range = config.object_range
+
+    object_reader = S3AsyncGetRangeRead(session, request_id,
```
Comment (mbcortx) [inline s3/replication/common/tests/system/test_range_read.py:59]: can we combine test for getobject and get range read

**Review by mbcortx — COMMENTED** [review]
I have couple of comments

File: s3/replication/common/src/s3replicationcommon/s3_range_read.py (hunk 3 above)
Comment (priyanka25081999) [reply s3/replication/common/src/s3replicationcommon/s3_range_read.py]: Done. Update the code of GetObject.

File: s3/replication/common/src/s3replicationcommon/s3_range_read.py (hunk 4 above)
Comment (priyanka25081999) [reply s3/replication/common/src/s3replicationcommon/s3_range_read.py]: yes. removed s3_range_read class as I have combined the range read function in GetObject.

File: s3/replication/common/src/s3replicationcommon/s3_range_read.py (hunk 6 above)
Comment (priyanka25081999) [reply s3/replication/common/src/s3replicationcommon/s3_range_read.py]: yes. done

File: s3/replication/common/tests/system/test_range_read.py (hunk 7 above)
Comment (priyanka25081999) [reply s3/replication/common/tests/system/test_range_read.py]: Yes we can. I have combined tests for getobject and get range read. 
done.

File: s3/replication/common/src/s3replicationcommon/s3_put_object_tagging.py (hunk 1 above)
Comment (priyanka25081999) [reply s3/replication/common/src/s3replicationcommon/s3_put_object_tagging.py:131]: okay.
Here, I am removing a temporary tagset file which is a duplicate of the original tagset file. 
Original object_tagset.xml file is still there in the config folder.

File: s3/replication/common/src/s3replicationcommon/s3_range_read.py (hunk 5 above)
Comment (priyanka25081999) [reply s3/replication/common/src/s3replicationcommon/s3_range_read.py]: done

**Review by mbcortx — APPROVED** [review]
LGTM

**Review by mbcortx — APPROVED** [review]

File: s3/replication/common/tests/system/config/config.yaml [hunk 8]
```diff
@@ -7,6 +7,8 @@ target_bucket_name: targetbucket # Buckets does exist on server
 object_name_prefix: foo_
 object_tag_name: user-tag
 object_tag_value: user-tag-value
+offset: -1
```
Comment (rajkumarpatel2602) [inline s3/replication/common/tests/system/config/config.yaml:10]: please make it a place holder.
e.g. <range-read-offset>, <range-read-length>
same for all places.

File: s3/replication/common/tests/system/test_get_object.py [hunk 9]
```diff
@@ -52,22 +52,35 @@ async def main():
...
 
     reader_generator = object_reader.fetch(object_size)
     async for _ in reader_generator:
         pass
 
-    if object_size == object_reader.get_content_length():
-        logger.info("Content-Length matched!")
-        logger.info("S3AsyncGetObject test passed!")
+    content_length = object_reader.get_content_length()
+
```
Comment (rajkumarpatel2602) [inline s3/replication/common/tests/system/test_get_object.py:69]: put comment on each case. 
e.g.
"range-read case"
"normal object read"

File: s3/replication/common/src/s3replicationcommon/aws_v4_signer.py (hunk 2 above)
Comment (rajkumarpatel2602) [reply s3/replication/common/src/s3replicationcommon/aws_v4_signer.py:199]: %s/obj_range/read_range/gc

File: s3/replication/common/src/s3replicationcommon/aws_v4_signer.py [hunk 10]
```diff
@@ -207,6 +208,8 @@ def prepare_signed_header(
         query_params - URL safe query param string,
                       example param1=abc&param2=somevalue
         body - content
+        Range - range of bytes string,
+                example bytes=start_range-end_range
```
Comment (rajkumarpatel2602) [inline s3/replication/common/src/s3replicationcommon/aws_v4_signer.py:212]: total_read_bytes = range_start - range_end

File: s3/replication/common/tests/system/config/config.yaml (hunk 8 above)
Comment (priyanka25081999) [reply s3/replication/common/tests/system/config/config.yaml]: done

File: s3/replication/common/tests/system/test_get_object.py (hunk 9 above)
Comment (priyanka25081999) [reply s3/replication/common/tests/system/test_get_object.py:69]: done

File: s3/replication/common/src/s3replicationcommon/aws_v4_signer.py (hunk 10 above)
Comment (priyanka25081999) [reply s3/replication/common/src/s3replicationcommon/aws_v4_signer.py]: done

**Review by mbcortx — APPROVED** [review]

**Comment by johnbent** [comment]
Hello @mbcortx @priyanka25081999 and @rajkumarpatel2602 . This is a great example of teamwork. Looks really good.

**Pull request merged**